File: python/html_pages_converter.py
"""
This is a variant on Luca Minudel's original exercise created by Emily Bache
"""
import html as html_converter


class HtmlPagesConverter:
    def __init__(self, filename):
        self.filename = filename
        self.breaks = [0]
        cumulative_char_count = 0
        for line in open(filename, "r"):
            cumulative_char_count += len(line) + 1  # add one for the newline
            if "PAGE_BREAK" in line:
                page_break_position = cumulative_char_count
                self.breaks.append(page_break_position)

    def get_html_page(self, page):
        html_page = ""
        f = open(self.filename, "r")
        f.seek(self.breaks[page])
        for line in f:
            if "PAGE_BREAK" in line:
                break
            line = line.rstrip()
            html_page += html_converter.escape(line, quote=True)
            html_page += "<br />"
        return html_page

File: swift/HtmlConverter/Sources/HtmlPagesConverter.swift
import Foundation

public class HtmlPagesConverter {
    private var _fileName: String
    private var breaks: [Int] = []

    public var fileName: String { _fileName }
    
    public init(fileName: String) throws {
        self._fileName = fileName

        breaks.append(0)
        let lines = AppFileReader(fileName: fileName)
            .readLines()

        var cumulativeCount = 0
        for line in lines {
            cumulativeCount += line.count + 1 // Add one for newline
            if line.contains("PAGE_BREAK") {
                let pageBreakPosition = cumulativeCount
                breaks.append(pageBreakPosition)
            }
        }
    }

    func getHtmlPage(_ page: Int) throws -> String {
        let lines = AppFileReader(fileName: _fileName)
            .readLines()
        let pageBreak = breaks[page]
        var htmlPage = ""

        var offset = 0
        for line in lines {
            offset += line.count

            // Skip lines until we're at the page's first line index
            guard offset >= pageBreak else { continue }

            if line.contains("PAGE_BREAK") {
                break
            }

            htmlPage += HttpUtility.htmlEncode(line)
            htmlPage += "<br />"
        }

        return htmlPage
    }
}
